import/index_java/src/org/vufind/index/CallNumberTools.java

```diff
@@ -40,34 +41,66 @@
 public class CallNumberTools
 {
     /**
-     * Extract the call number label from a record
+     * Extract the call number labels from a record
      * @param record MARC record
-     * @return Call number label
+     * @return Call number labels
      */
-    public String getCallNumberLabel(final Record record) {
+    public List<String> getCallNumberLabel(final Record record) {
 
-        return getCallNumberLabel(record, "090a:050a");
+        return getCallNumberLabel(record, "050a:082a:083a:090a");
     }
 
     /**
-     * Extract the call number label from a record
+     * Extract the call number labels from a record.
+     * fieldSpec might include 952e for FOLIO, in which case we do not know the type of call number
+     * (also sometimes call numbers are in the wrong MARC field for their type).
+     * If LCC: classLetters, classLetters + classDigits, classLetters + classDigits + classDecimal
+     * If Dewey: classDigits, classDigits + classDecimal
+     * Otherwise: everything up until the dot (or everything if there is no dot)
      * @param record MARC record
      * @param fieldSpec taglist for call number fields
-     * @return Call number label
+     * @return Call number labels
      */
-    public String getCallNumberLabel(final Record record, String fieldSpec) {
-
-        String val = SolrIndexer.instance().getFirstFieldVal(record, fieldSpec);
-
-        if (val != null) {
-            int dotPos = val.indexOf(".");
-            if (dotPos > 0) {
-                val = val.substring(0, dotPos);
+    public List<String> getCallNumberLabel(final Record record, String fieldSpec) {
```

Comment:
For backward-compatibility, how about renaming this method to getCallNumberLabels, and then keeping a getCallNumberLabel method that returns the first value from getCallNumberLabels?

Reply:
I thought of doing that, but I wasn't sure what solution you would prefer (more methods - 2 are needed, forcing a config change or keeping the old name without an s). I opted for the last one, but I don't mind changing it. I will add 2 methods and deprecate the old ones. But I guess I have to return an array, and it makes more sense to me to use all values - this way people don't have do change anything to benefit from the change.

Reply:
I think we need to pick an approach and go all the way in that direction: either abandon the old methods entirely (which was your initial approach) or else maintain true backward compatibility (keep the same return value and behavior of the legacy methods -- i.e. return just one String). My thinking is that if somebody somewhere is using the existing method to populate a single-valued field, making it suddenly multi-valued will break indexing. That's why I proposed keeping it as deprecated but maintaining existing behavior. However, I also realize that this situation is probably highly unlikely, which might be an argument toward abandoning the legacy method entirely.

I don't really have strong feelings about this, but I think keeping the old method name and changing the behavior is the "worst of both worlds": it creates a proliferation of methods without actually maintaining true backward compatibility.

Does that make sense? What do you think? :-)

Reply:
Are you sure it will not break import with a multi-valued field in Solr ?

Reply:
Passing a single value to a multi-valued field in SolrMarc should not cause a problem. But of course passing multiple values to a single-valued field definitely will.

Reply:
Actually, this is not going to work with the new implementation (it would return a 2 letters call number label). So I will use the old implementation for these methods instead.
Chances are some people will not benefit from the improvements because they will not update their config file - so we should suggest the change as an upgrade step.

Reply:
If people are following best practices and only overriding custom import rules in marc_local.properties instead of copying the whole marc.properties, then they should get the enhancement automatically. I think it's fairly unlikely that people will currently have custom local callnumber-label settings, though we should definitely make note of this in the changelog to be on the safe side (I'll put a TODO checkbox on this PR as a reminder).

Reply:
FOLIO users using OAI/PMH should customize it (to use 952e).